### PR TITLE
Inequality constraints on columns

### PR DESCRIPTION
**Pull request by matbesancon**
Adapting the interface from interval constraints for one-sided inequalities.

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -47,14 +47,57 @@ function MOI.add_constrained_variable(o::Optimizer, set::S) where {S <: MOI.Inte
...
+    return MOI.SingleVariable(MOI.VariableIndex(ci.value))
+end
+
+function MOI.get(o::Optimizer, ::MOI.ConstraintSet, ci::MOI.ConstraintIndex{MOI.SingleVariable, MOI.Interval{Float64}})
+    var_idx = Cint(ci.value)
+    num_col = Ref{Cint}(0)
+    num_nz = Ref{Cint}(0)
+    lower = Ref{Cdouble}()
+    upper = Ref{Cdouble}()
+    _ = CWrapper.Highs_getColsByRange(o.model.inner, var_idx, var_idx, num_col, C_NULL, lower, upper, num_nz, C_NULL, C_NULL, C_NULL)
```
Comment (mlubin) [inline src/MOI_wrapper.jl:68]: Shouldn't we check the return of the function for errors?

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -47,14 +47,57 @@ function MOI.add_constrained_variable(o::Optimizer, set::S) where {S <: MOI.Inte
...
+    num_col[] == 1 || error("Unexpected HiGHS state, number of columns should be one for valid range, is $(num_col[])")
+    return MOI.Interval(lower[], upper[])
```
Comment (mlubin) [inline src/MOI_wrapper.jl:70]: It should be an error if the user queries a property of a constraint that wasn't added, e.g., if they added a `LessThan` and a `GreaterThan` constraint and there was never an `Interval` in the model added, this behavior isn't correct.

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -47,14 +47,57 @@ function MOI.add_constrained_variable(o::Optimizer, set::S) where {S <: MOI.Inte
...
+end
+
+function MOI.add_constraint(o::Optimizer, sg::MOI.SingleVariable, set::MOI.LessThan)
+    ci_interval = MOI.ConstraintIndex{MOI.SingleVariable, MOI.Interval{Float64}}(sg.variable.value)
+    interval_set = MOI.get(o, MOI.ConstraintSet(), ci_interval)
+    if set.upper <= interval_set.upper
+        _ = CWrapper.Highs_changeColBounds(o.model.inner, Cint(sg.variable.value), Cdouble(interval_set.lower), Cdouble(set.upper))
+    end
+    return MOI.ConstraintIndex{MOI.SingleVariable, MOI.LessThan{Float64}}(sg.variable.value)
```
Comment (mlubin) [inline src/MOI_wrapper.jl:79]: Likewise, if the user never added a `LessThan` constraint for this variable, this function should return an error.

**Comment by matbesancon** [comment]
> It's probably simpler to drop support for SingleVariable-in-Interval

I would agree for most solver types but HiGHS native constraints are intervals on variables and on linear expressions, so the first thing it supports is that, and then we need to adapt some additional indirection for single-sided inequalities

File: src/MOI_wrapper.jl (hunk 2 above)
Comment (matbesancon) [reply src/MOI_wrapper.jl:70]: what do you think is the best approach here, because there is always a value (even if it's +inf)

**Comment by matbesancon** [comment]
> Likewise, if the user never added a LessThan constraint for this variable, this function should return an error.

There is a conceptual problem with this --> do we consider that adding Interval(a,b) is equivalent to adding GreaterThan(a) and LessThan(b), if yes that's fine, we only error if we are about to return a +- Inf.

If we should have an error if getting a LessThan when we never added a corresponding LessThan (and only less than), then we need to track what has been added or not, which is much of a burden

**Comment by mlubin** [comment]
> do we consider that adding Interval(a,b) is equivalent to adding GreaterThan(a) and LessThan(b)

No, we don't, at least that's the intention.

> If we should have an error if getting a LessThan when we never added a corresponding LessThan (and only less than), then we need to track what has been added or not, which is much of a burden

Yes, that should be tracked. This is necessary for correctly implementing `ListOfConstraintIndices`. Depending on the order of the tests, I think [this test](https://github.com/JuliaOpt/MathOptInterface.jl/blob/2a8603407fdabce4f322cd90fe646f3a0efdc6a6/src/Test/UnitTests/basic_constraint_tests.jl#L267) should fail if the solver interface confuses intervals with GreaterThan and LessThan.

**Comment by matbesancon** [comment]
this is fairly heavy as an abstraction, it feels like all solvers have to manually keep track of every constraint in the model

**Comment by mlubin** [comment]
Not exactly every constraint added. GLPK keeps a flag for each variable so it knows which kinds of MOI constraints the variable bounds correspond to: https://github.com/JuliaOpt/GLPK.jl/blob/ffd27144055c9ffa4a6a32223cb0b8e9fbd95084/src/MOI_wrapper.jl#L7. For rows of the constraint matrix it also stores the corresponding MOI set. This is the price for having multiple ways to express a constraint in MOI that map to a single representation in the solver.

**Comment by matbesancon** [comment]
>  keeps a flag for each variable so it knows which kinds of MOI constraints the variable bounds correspond to

Ok but if you successively add a LessThan, and then a GreaterThan, what flag should be used, GreaterThan / LessThan or interval? And should the new bound override the previous?

**Comment by mlubin** [comment]
> Ok but if you successively add a LessThan, and then a GreaterThan, what flag should be used, GreaterThan / LessThan or interval?

In the GLPK case, `LESS_AND_GREATER_THAN`. They're really separate constraints and should never be confused by the interface.

> And should the new bound override the previous?

No, adding more than one constraint that maps to a variable lower or upper bound is an error. See https://www.juliaopt.org/MathOptInterface.jl/dev/apireference/#MathOptInterface.add_constraint, https://www.juliaopt.org/MathOptInterface.jl/dev/apireference/#MathOptInterface.LowerBoundAlreadySet, https://www.juliaopt.org/MathOptInterface.jl/dev/apireference/#MathOptInterface.UpperBoundAlreadySet.

**Comment by odow** [comment]
Closing because I have a different PR in the works.

**Pull request closed without merging**